File: 0846-hand-of-straights/0846-hand-of-straights.java
class Solution {
    public boolean isNStraightHand(int[] hand, int groupSize) {
        int n = hand.length;
        if (n % groupSize != 0) {
            return false;
        }
        
        TreeMap<Integer, Integer> count = new TreeMap<>();
        for (int card : hand) {
            count.put(card, count.getOrDefault(card, 0) + 1);
        }
        
        while (!count.isEmpty()) {
            int first = count.firstKey();
            int freq = count.get(first);
            
            for (int i = first; i < first + groupSize; i++) {
                if (!count.containsKey(i)) {
                    return false;
                }
                
                int newFreq = count.get(i) - freq;
                if (newFreq < 0) {
                    return false;
                } else if (newFreq == 0) {
                    count.remove(i);
                } else {
                    count.put(i, newFreq);
                }
            }
        }
        
        return true;
    }
}

File: 0846-hand-of-straights/0846-hand-of-straights.cpp
class Solution {
public:
    bool isNStraightHand(vector<int>& hand, int groupSize) {
        int n = hand.size();
        if (n % groupSize != 0) {
            return false;
        }
        
        map<int, int> count;
        for (int card : hand) {
            count[card]++;
        }
        
        while (!count.empty()) {
            int first = count.begin()->first;
            int freq = count[first];
            
            for (int i = first; i < first + groupSize; i++) {
                if (count.find(i) == count.end()) {
                    return false;
                }
                
                count[i] -= freq;
                if (count[i] < 0) {
                    return false;
                } else if (count[i] == 0) {
                    count.erase(i);
                }
            }
        }
        
        return true;
    }
};